doc/neps/nep-0049.rst

```diff
@@ -118,20 +118,20 @@ NumPy C-API functions
 
         typedef struct {
             char name[128];  /* multiple of 64 to keep the struct aligned */
-            PyDataMem_AllocFunc *alloc;
-            PyDataMem_ZeroedAllocFunc *zeroed_alloc;
-            PyDataMem_FreeFunc *free;
-            PyDataMem_ReallocFunc *realloc;
+            PyDataMemAllocator allocator; 
         } PyDataMem_Handler;
 
-    where the function's signatures are
+    where the allocator structure is
 
     .. code-block:: c
-
-        typedef void *(PyDataMem_AllocFunc)(size_t size);
-        typedef void *(PyDataMem_ZeroedAllocFunc)(size_t nelems, size_t elsize);
-        typedef void (PyDataMem_FreeFunc)(void *ptr, size_t size);
-        typedef void *(PyDataMem_ReallocFunc)(void *ptr, size_t size);
+        /* The declaration of free differs from PyMemAllocatorEx */ 
+        typedef struct {
+            void *ctx;
+            void* (*malloc) (void *ctx, size_t size);
+            void* (*calloc) (void *ctx, size_t nelem, size_t elsize);
+            void* (*realloc) (void *ctx, void *ptr, size_t new_size);
+            void (*free) (void *ctx, void *ptr, size_t size);
```

Comment:
Out of curiosity, why is a size is passed to `free`.

Reply:
This allows a strategy where, like in [`np_cache_free`](https://github.com/numpy/numpy/blob/e69faef03428b82f5651a5fb521610ddea2bd22a/numpy/core/src/multiarray/alloc.c#L161) the memory is not actually returned to the operating system rather is kept for a possible future allocation of the same number of bytes. It is necessary for the current, default, allocation strategy.

Reply:
If this isn't already mentioned in the NEP somewhere, can we include the justification of having this extra this parameter (i'm not opposed to it, but I think it's unusual enough to be worth explaining).

Reply:
A `mmap`/`munmap`-based allocation policy could also benefit from the `size` argument in `free`.

Reply:
It would be good know how it is intended to be used, and how, or where, it is tracked. Can it differ from the allocation size?

Reply:
It shouldn't differ. However, the `shift_allocator` test that examines exactly this behavior, reports 3 unmatched sizes on exit (GC).

While there are always workarounds, two scenarios that would benefit from `size` on `free` are:
1. `mmap`/`munmap`-based allocators
2. Pooled allocators / Memory arenas

Workarounds (in case that we decide to stick with `PyMemAllocatorEx`) include:
1. Hiding the size in the preamble (like the `shift_allocator` test does)
2. Implementing a `ptr`-`size` dictionary inside the `ctx`

However, there are cases where even the first could be considered an overkill, e.g. memory-mapped page-aligned allocators (`mmap` always returns page-aligned addresses, which means it would require a wasted page to just store the `size`).

Reply:
If we move to `PyMemAllocatorEx` we will need to do the second work-around for the current, default allocator since it uses `npy_free_cache(void * p, npy_uintp sz)`, where `sz` is something like `PyArray_NBYTES(self)`

Reply:
As long as the `PyDataMem_UserFREE` function signature remains [as is](https://github.com/mattip/numpy/blob/b4db79f2e287e41df1e98d895d61822ab809d808/numpy/core/src/multiarray/alloc.c#L402-L406), and since already the default allocator is "privileged" (e.g. `npy_free_cache` doesn't even accept a `ctx`, but also integrates `tracemalloc`/`eventhook` logic), I think there won't be any problem with it.

It's our decision whether or not `size` will be also advertised to custom allocators on `free` (#L408). But definitely the fact that the `default_allocator` needs it, makes a strong argument in favor of doing so.